mooncake-integration/store/store_py.cpp

```diff
@@ -725,6 +725,158 @@ int DistributedObjectStore::get_into(const std::string &key, void *buffer,
+            void *chunk_ptr = static_cast<char *>(buffer) + offset;
+            key_slices.emplace_back(Slice{chunk_ptr, chunk_size});
+            offset += chunk_size;
+        }
+        all_slices[key] = key_slices;
+    }
+
+    ErrorCode batch_put_err = client_->BatchPut(keys, all_slices, config);
+
+    std::vector<int> results(keys.size());
+    if (batch_put_err != ErrorCode::OK) {
+        LOG(ERROR) << "BatchPut failed with error: " << toString(batch_put_err);
+        std::fill(results.begin(), results.end(), toInt(batch_put_err));
+    } else {
+        std::fill(results.begin(), results.end(), 0);
+    }
+
+    return results;
+}
+
+std::vector<int> DistributedObjectStore::batch_get_into(
+    const std::vector<std::string> &keys, const std::vector<void *> &buffers,
+    const std::vector<size_t> &sizes) {
+    auto start_time = std::chrono::steady_clock::now();
+    if (!client_) {
+        LOG(ERROR) << "Client is not initialized";
+        return std::vector<int>(keys.size(), -1);
+    }
+
+    if (keys.size() != buffers.size() || keys.size() != sizes.size()) {
+        LOG(ERROR) << "Mismatched sizes for keys, buffers, and sizes";
+        return std::vector<int>(keys.size(), -1);
+    }
+
+    std::vector<int> results(keys.size());
+    mooncake::Client::BatchObjectInfo
+        object_infos;  // This is BatchGetReplicaListResponse
+
+    // Step 1: Batch query object info
+    ErrorCode batch_query_err = client_->BatchQuery(keys, object_infos);
+    if (batch_query_err != ErrorCode::OK) {
+        LOG(ERROR) << "BatchQuery failed with error: "
+                   << toString(batch_query_err);
+        std::fill(results.begin(), results.end(), toInt(batch_query_err));
+        return results;
+    }
+
+    // Step 2: Prepare slices for each key
+    std::unordered_map<std::string, std::vector<mooncake::Slice>> all_slices;
+    std::vector<std::string> valid_keys;
+    for (size_t i = 0; i < keys.size(); ++i) {
+        const auto &key = keys[i];
+        auto it = object_infos.batch_replica_list.find(key);
+
+        if (it == object_infos.batch_replica_list.end()) {
+            results[i] = -toInt(ErrorCode::OBJECT_NOT_FOUND);
+            continue;
+        }
+
+        auto &replica_list = it->second;
+        if (replica_list.empty()) {
+            LOG(ERROR) << "Internal error: replica_list is empty for key: "
+                       << key;
+            results[i] = -1;
+            continue;
+        }
+
+        auto &replica = replica_list[0];
+        uint64_t total_size = 0;
+        for (auto &handle : replica.buffer_descriptors) {
+            total_size += handle.size_;
+        }
+
+        if (sizes[i] < total_size) {
+            LOG(ERROR) << "User buffer too small for key: " << key
+                       << ". Required: " << total_size
+                       << ", provided: " << sizes[i];
+            results[i] = -1;
+            continue;
```

Comment:
Do we need to find out the longest prefix where all the keys exist? I have heard that the caller may only be interested in consecutive prefixes.

Reply:
Yep, you got it right.